rfd/0107-desktop-access-performance-research.md

```diff
@@ -0,0 +1,95 @@
+
+```
+| message type (30) | cache_id uint32 | cache_index uint32 | top uint32 | left uint32 | right uint32 | bottom uint32 |
+```
+
+Update to the UI is also required. The browser memory and javascript's array of hash tables, where cache id is the index of the array, and the cache index is a key in the hash table, will be used to store bitmap data.
+
+
+There's a sequence diagram which shows the flow of the messages between all components:
+
+```mermaid
+sequenceDiagram
+    Windows Server->>Teleport Windows Desktop Service: rdp: store bitmap
+    Teleport Windows Desktop Service->>Teleport Proxy: tdp: store bitmap
+    Teleport Proxy->>Web Browser: tdp: store bitmap
+    Web Browser->>cache in web browser memory: store bitmap
+    Windows Server->>Teleport Windows Desktop Service: rdp: load bitmap
+    Teleport Windows Desktop Service->>Teleport Proxy: tdp: load bitmap
+    Teleport Proxy->>Web Browser: tdp: load bitmap
+    cache in web browser memory->>Web Browser: load bitmap
+```
+
+- server sends [message](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/3c76f8fb-fcaa-4ca6-9d2e-3f449bdbed40) to inform client to store bitmap in the cache
+- windows destkop service translates rdp message into tdp message to store bitmap and sends it to the proxy
+- proxy forwards message to web browser
+- web browser stores bitmap in the in-memory cache
+- server sends [message](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/84c2ec2f-f776-405b-9b48-6894a28b1b14) to inform client to load bitmap from the cache
+- windows destkop service translates rdp message into tdp message to load bitmap and sends it to the proxy
+- proxy forwards message to web browser
+- web browser loads bitmap from the in-memory cache
+
+#### session recordings
+Since we no longer send every bitmap over the wire and to keep session recordings to work we'll also need to keep the cache of the bitmaps at the proxy. When we encounter the `save bitmap message` we'll need to store that bitmap at the proxy and then whenver `load bitmap message` will be sent we'll need to load bitmap from the cache and generate appropriate event with the bitmap data. 
+
+
+### Process bitmaps in Rust library
+
+While interacting with the remote desktop using the RDP protocol, most of the protocol messages exchanged between the server and a client are related to rendering bitmaps. Messages almost always contain compressed data to reduce bandwidth usage and latency. Rendering compressed bitmaps on the screen requires uncompressing data first using decompress algorithm.
+ It is the most common operation while using the protocol, so it must execute extermaly fast to provide good user experience.
+The best performance can be achived be moving the bitmap decompression procdure from the Go client to the faster compiled library written in the Rust language. It'll also allow to simplyfy the way we encode and decode PNG tdp messages.
```

Comment:
I thought decompression was already done in Rust. Can you link to the Go code you're speaking about here?

Reply:
What I really should've done is to add annotation that this is already implemented. My intention was to include it in in the RFD so that if anyone else is looking at this RFD will see what was important for us when we were checking for performance issues in the Desktop Access. I think this was really important so just wanted to include it here.

Reply:
Decompression has always been done in Rust. Do you mean PNG encoding?

Reply:
Yes! For some reason I mixed these two terms. I'll revisit the doc and make sure that I'm using the right terms 😄